src/github.com/getlantern/detour/detour_conn.go

```diff
@@ -0,0 +1,82 @@
+package detour
+
+import (
+	"io"
+	"net"
+	"sync/atomic"
+)
+
+type detourConn struct {
+	net.Conn
+	addr string
+	// keep track of the total bytes read in this connection
+	readBytes uint64
+
+	// 1 == true, 0 == false, atomic
+	errorEncountered uint32
+	// a flag telling if connection is closed
+	closed uint32
+}
+
+func dialDetour(network string, addr string, dialer dialFunc, ch chan conn) {
+	go func() {
+		log.Tracef("Dialing detour connection to %s", addr)
+		conn, err := dialer(network, addr)
+		if err != nil {
+			log.Errorf("Dial detour to %s failed: %s", addr, err)
+			return
+		}
+		log.Tracef("Dial detour to %s succeeded", addr)
+		ch <- &detourConn{Conn: conn, addr: addr, readBytes: 0}
+	}()
+	return
+}
+
+func (dc *detourConn) ConnType() connType {
+	return connTypeDetour
+}
+
+func (dc *detourConn) FirstRead(b []byte, ch chan ioResult) {
+	dc.doRead(b, ch)
+}
+
+func (dc *detourConn) FollowupRead(b []byte, ch chan ioResult) {
+	dc.doRead(b, ch)
+}
+
+func (dc *detourConn) doRead(b []byte, ch chan ioResult) {
+	go func() {
+		n, err := dc.Conn.Read(b)
+		atomic.AddUint64(&dc.readBytes, uint64(n))
+		defer func() { ch <- ioResult{n, err, dc} }()
+		if err != nil {
+			if err != io.EOF {
+				atomic.AddUint32(&dc.errorEncountered, 1)
+			}
+			return
+		}
+	}()
+	return
+}
+
+func (dc *detourConn) Write(b []byte, ch chan ioResult) {
+	go func() {
+		n, err := dc.Conn.Write(b)
+		defer func() { ch <- ioResult{n, err, dc} }()
```

Comment:
The function is already about to end at this point, I think the defer is not necessary.